### PR TITLE
doc(angular): add Angular to Aurelia migration

### PR DESCRIPTION
**Pull request by jedd-ahyoung**
Add documentation for migrating an Angular 1.x app to Aurelia.

### CODE REVIEW

**Comment by jedd-ahyoung** [comment]
This is not the best PR, as the documentation is still a work in progress. However, I'd like to get this out as soon as possible to begin obtaining community support. I believe that the foundation is here, and I think that we can flesh this out as the community asks questions (especially regarding custom elements, custom attributes, and scope questions).

**Comment by EisenbergEffect** [comment]
Awesome! This is probably good enough to release as an initial version. I can do a quick cleanup and make sure the doc format is correct. Do you have any objections to me adding this to the online docs?

**Comment by jedd-ahyoung** [comment]
@EisenbergEffect I don't. My hope is that we can get some community
feedback and make it better over time. Go ahead and add it. I'll continue
to work on it when I can.

2016-03-24 6:40 GMT-04:00 Rob Eisenberg <notifications@github.com>:

File: doc/article/en-US/angular1-to-aurelia.md [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+	</source-code>
+</code-listing>
+
+Aurelia handles things a little differently. Instead of framework-specific modules, Aurelia leverages native Javascript modules using either the ES6 import statement or the `require`-style AMD syntax. (These end up transpiled to AMD modules for browser use.) When registering objects with Angular modules Angular made a distinction between `service`, `factory`,  `constant`, and `value` - all syntactic sugar for `provider`. Aurelia, in contrast, makes no special framework distinction between "services", "factories", and other module types - any viewmodels or services are just simple Javascript modules, simplifying the way you can write your code.
+
+So, the angular code above might end up looking more like this:
+
+<code-listing heading="Aurelia Component Definitions">
+	<source-code lang="ES 2015/2016">@inject(Q, HttpClient)
+		export class myService {
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: forgot to inject dependencies?

Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: You haven't added imports for dependencies not defined in this file - why, is it conscious decision?

File: doc/article/en-US/angular1-to-aurelia.md [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+			}
+		}
+		
+		export var myValue = value;
+	</source-code>
+</code-listing>
+
+As long as you define your exports semantically, Aurelia doesn't need to register or define them in a specific way ("factory" vs "service" vs "value"). Everything is an exported module; everything is generic, and modules can be imported and then injected using Aurelia's dependency injection.
+
+For more information on dependency injection, see the documentation here.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: missing link to docs?

File: doc/article/en-US/angular1-to-aurelia.md [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+	<source-code lang="ES 2015/2016">
+		function Users (httpClient) {
+			...
+		}
+	</source-code>
+</code-listing>
+
+> Note: Dependency Injection
+> Angular was able to inject Angular-specific dependencies based on the argument name (such as `$scope` and `$http`) through its service locator. Because Aurelia does not couple your application to Aurelia-specific implementations, all dependencies must be explicitly injected.
+
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: It might be just me, but i'm afraid this example and comment 
> Because Aurelia does not couple your application to Aurelia-specific implementations, all dependencies must be explicitly injected.

doesn't help much to understand how to create viewmodel, that will get its dependencies. I'd add a class with inject annotation instead of this example (as it has a dependency, that needs to be injected).

File: doc/article/en-US/angular1-to-aurelia.md [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+#### Routing
+
+Aurelia supports client-side routing in the browser. The main difference between Angular's router and Aurelia's router is in the configuration; the Aurelia router can be configured outside of startup, and it can be configured asynchronously. Finally, the Aurelia router supports nested routers (also known as child routers) as well as multiple viewports.
+
+An Angular router configuration, in startup, may look like this:
+
+Alternatively, an Aurelia router is configured inside the viewmodel that uses it. Thus, Aurelia configuration may look like this:
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: > inside the viewmodel

I think viewmodel is not correct term here - I would have used "aurelia configuration class" here, but but maybe someone can suggest even better term.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+		
+		    this.router = router;
+		  }
+		}
+	</source-code>
+</code-listing>
+
+> Note: Router Migration
+> Migration to Aurelia will be easier if your application uses Angular's new router; this router shares many similarities with the router that Aurelia uses, and the configuration will be easier to migrate.
+
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: @EisenbergEffect, would it be too much to montion that the mastermind of Aurelia also created Angular's new router: https://github.com/angular/router/graphs/contributors :P

File: doc/article/en-US/angular1-to-aurelia.md [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+#### Removing `$scope`
+
+Angular made heavy use of an Angular-specific service named `$scope`; responsible for computed properties, change detection (through `$scope.watch()`), signalling (via `$scope.broadcast` and `$scope.emit`), and view-to-viewmodel binding, this service was often a staple of an Angular controller.
+
+A typical controller might yield something like this:
+
+<code-listing heading="Angular Controller Configuration">
+	<source-code lang="ES 2015/2016">
+		angular.controller('Users', ['$scope', '$http', function ($scope, $http) {
+			$scope.heading = 'Welcome to the Angular Navigation App!';	$scope.firstName = 'John';
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: missing new line

File: doc/article/en-US/angular1-to-aurelia.md [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+		}]);
+	</source-code>
+</code-listing>
+
+In Aurelia, viewmodel properties are declared as properties on the viewmodel object.
+
+The Aurelia version of our typical Angular controller might look like this:
+
+<code-listing heading="Aurelia Controller Configuration">
+	<source-code lang="ES 2015/2016">		export class Welcome {
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: missing new line?

File: doc/article/en-US/angular1-to-aurelia.md [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+		  submit() {
+		    this.previousValue = this.fullName;
+		    alert(`Welcome, ${this.fullName}!`);
+		  }
+		}
+	</source-code>
+</code-listing>
+
+> Note: Controller Migration
+> Later versions of Angular supported the "controller as" syntax, allowing developers to declare viewmodel properties on the `this` property. However, `$scope` was still necessary for watches and signalling.) Your application will be easier to migrate to Aurelia if you are only using `$scope` for signalling.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: I'd add an example for "controller as" syntax as well before showing corresponding Aurelia controller.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+
+### Migrating your Views
+
+Fourth step, part two - we'll migrate the views. Aurelia views and Angular views are very similar. However, Aurelia views follow the Web Components specification, requiring views to be wrapped in a `<template>` tag. Further, references to custom views or custom elements must be included in the view using the `<require>` tag.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: > references to custom views or custom elements must be included in the view using the `<require>` tag.

...unless they are made globally available using `aurelia.use.globalResources(..)`

File: doc/article/en-US/angular1-to-aurelia.md [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+
+#### Naming Conventions
+
+In Angular, a view was often strongly coupled to a viewmodel (or controller) instance, using something similar to the following code:
+
+<code-listing heading="Angular View">
+	<source-code lang="ES 2015/2016">		<section data-ng-controller="myController as MyCtrl">
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: missing new line?

File: doc/article/en-US/angular1-to-aurelia.md [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+		...
+		</section>
+	</source-code>
+</code-listing>
+
+Aurelia, by default, prefers convention over configuration. Aurelia's default convention is that a viewmodel will have a view with the same name. To take advantage of this, keep a view in the same directory alongside its viewmodel, and make sure that the HTML file and Javascript file share the same filename.
+
+> Note: Convention vs configuration
+> This behavior is also configurable using the ViewLocator service. By configuring Aurelia's ViewLocator, you can override Aurelia's defaults. Views and viewmodels can be linked by location, by a file naming convention, or by other means.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: I think mentioning `@useView('./some-other-file-name.html')` would be much more worth compared to ViewLocator (i'd just mention it as an advanced option or maybe even leave it out from here).

File: doc/article/en-US/angular1-to-aurelia.md [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+
+Aurelia makes use of ES6 template strings in its templates, so the Aurelia version becomes this:
+
+<code-listing heading="Aurelia Interpolation">
+	<source-code lang="HTML">
+		<div id="hello">Hello, my name is ${name}!</div>
+	</source-code>
+</code-listing>
+
+Aurelia's template strings use pipe syntax for ValueConverters (filters in Angular).
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: I'd rephrase this sentence, so be more clear, that the syntax is same for filters in Angular that correspond to ValueConverters in Aurelia

File: doc/article/en-US/angular1-to-aurelia.md [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+
+#### Template behaviors
+
+`ng-show` becomes `show.bind`. `ng-if` becomes `if.bind`. `ng-click` becomes `click.delegate`.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Probably it is worth explaining how similarly to `click.delegate` you could use delegate for all other events as well (maybe even mention event-related attribute names without "on" prefix)

File: doc/article/en-US/angular1-to-aurelia.md [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+
+<code-listing heading="Angular Interpolation">
+	<source-code lang="HTML">
+		<ul data-ng-repeat="item in items">
+			<li id="{{item.id}}">{{item.name}}</li>
+		</ul>
+	</source-code>
+</code-listing>
+
+In Aurelia, the repeat is not on the parent element. It looks like this:
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: @EisenbergEffect, i think it would be worth explaining why different approach was chosen - I guess Your comments would be very much appreciated here to avoid frustrations :)

File: doc/article/en-US/angular1-to-aurelia.md [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+Instead of a configuration object, Aurelia takes a class instance.
+
+An Angular directive may have looked like this:
+
+<code-listing heading="Angular Custom Components">
+	<source-code lang="ES 2015/2016">
+		angular.directive('display', ['$window', '$document', 'animate', function ($window, $document, animate) {
+			return {
+				scope: true,
+				restrict: 'A',
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: You talk here about Custom elements, but `display` directive is restricted to 'A' (attribute), that corresponds to custom attributes

File: doc/article/en-US/angular1-to-aurelia.md [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+				}
+			};
+		}])
+	</source-code>
+</code-listing>
+
+Converting this directive to an Aurelia custom element requires us to remove the scope references, migrate from `link` and `controller` properties to lifecycle methods, and simplify our object. Finally, we'll have to use Aurelia's decorator functions to register our object and its configuration as a custom element instance.
+
+<code-listing heading="Aurelia Custom Components">
+	<source-code lang="ES 2015/2016">		export class Display {
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: missing new line? Missing inject.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 17]
```diff
@@ -0,0 +1,526 @@
...
+			constructor (window, document, animate) {
+				this.window = window;
+				this.document = document;
+				this.animate = animate;
+			}
+			
+			attached () {
+
+			}
+		}
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: **Serious deficiency**. You should make the Aurelia class as similar to Angular directive. It is not fair to pick not the most trivial Angular directive as an example and show just a class constructor for Aurelia class that won't even work as custom element (not to mention functioning similarly).

File: doc/article/en-US/angular1-to-aurelia.md [hunk 18]
```diff
@@ -0,0 +1,526 @@
...
+
+Creating a custom element in Angular required us to write an Angular directive; we would return a configuration object that told Angular how to create our object through object properties including `scope`, `restrict`, `template`, `templateUrl`, and `link`.
+
+Instead of a configuration object, Aurelia takes a class instance.
+
+An Angular directive may have looked like this:
+
+<code-listing heading="Angular Custom Components">
+	<source-code lang="ES 2015/2016">
+		angular.directive('display', ['$window', '$document', 'animate', function ($window, $document, animate) {
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: I think there should be several examples of Angular element type directives to explain different ways that Angular directives could have been used and how to map them to Aurelia Custom element:
For example scope - isolated or not. For the first example I think the directive should have isolated scope (and probably accept different types of arguments from the caller, including primitives, objects, function callbacks with arguments)

File: doc/article/en-US/angular1-to-aurelia.md [hunk 19]
```diff
@@ -0,0 +1,526 @@
...
+			
+			attached () {
+
+			}
+		}
+	</source-code>
+</code-listing>
+
+> Warning: Scope
+> Custom elements in Aurelia cannot access the outer scope. This means that all custom elements in Aurelia essentially behave as though they had the Angular property `scope: { ... }` - custom elements have an isolated scope.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: I'd rather add initial example with isolated scope and another example where scope is inherited.

**Comment by EisenbergEffect** [comment]
@atsu85 Thanks for all the great feedback. This was a first draft by @jedd-ahyoung just to get something in place with the hopes that the community could fill it out more. Let me get this merged today or tomorrow. After that maybe you can submit some PRs on the doc to start improving the content?

File: doc/article/en-US/angular1-to-aurelia.md [hunk 20]
```diff
@@ -0,0 +1,526 @@
...
+
+#### Converting an Angular directive to an Aurelia custom attribute
+
+Custom attributes are handled similarly to custom elements in Aurelia. Custom attributes also use an isolate scope - they cannot access the outer scope.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: > Custom attributes are ...

vs 
"Angular directives restricted attributes (`restrict: 'A'`) are ..."

Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Example needed here.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 21]
```diff
@@ -0,0 +1,526 @@
...
+
+#### Transclusion
+
+Aurelia provides the `content` tag to provide transclusion features. This is part of the view itself, instead of being part of the directive.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Example needed here.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 22]
```diff
@@ -0,0 +1,526 @@
...
+
+## [Quick Reference](aurelia-doc://section/4/version/1.0.0)
+
+#### $scope
+
+No equivalent exists in Aurelia.
+
+Instead of `$scope.property`, use `this.property` inside of a viewmodel object.
+
+Instead of `$scope.$watch()`, use a computed property - or, to trigger a function, use the BindingEngine.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: > or, to trigger a function, use the BindingEngine

With my experience, this is rarely needed or reasonable (unless writing plugins, where this is much more often handy). It might be worth to explain how to get rid of `$scope.$watch` that is used in Angular app for different purposes (mostly as a workaround?).

File: doc/article/en-US/angular1-to-aurelia.md [hunk 23]
```diff
@@ -0,0 +1,526 @@
...
+
+Instead of `$scope.$emit()` and `$scope.$broadcast()`, use `EventAggregator.publish()`. Likewise, instead of $scope.$on, use `EventAggregator.subscribe()`.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Explain that with Angular has two functions to publish events (`$emit` dispatches events for components upwards to component hierarchy tree and `$broadcast` downwards). In Aurelia `EventAggregator.publish()` works pretty much like `$rootScope.$broadcast`, because everyone (not just viewmodels, custom elements/attributes) can choose to subscribe to listen the events.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 24]
```diff
@@ -0,0 +1,526 @@
...
+
+For signalling (using `$emit` and `$broadcast`, use Aurelia's event aggregator. For state that persists across viewmodels, use a separate service.
+
+#### $timeout
+
+No equivalent exists in Aurelia; it's not needed as Aurelia's change detection works differently to change detection in Angular. Aurelia will detect changes made from `setTimeout`, so use that.
+
+#### $cacheFactory
+
+Consider using Aurelia's cache plugin.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Add a link to the plugin.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 25]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $animate/$animateCss
+
+Use one of the 'aurelia-animate' plugins.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Add a link to the plugin.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 26]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $controller
+
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: What is planned here?

File: doc/article/en-US/angular1-to-aurelia.md [hunk 27]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $document
+
+Use `window.document`.
+
+#### $filter
+
+Use a `ValueConverter`, providing your own `toView` and `fromView` methods.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: It is worth mentioning, that unlike Angular filters, Aurelia ValueConverters are two-way. The `toView` method would contain code taken from Angular filter. Explain, that `fromView` can be used to push changes from view back to the viewmodel.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 28]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $http/$httpBackend
+
+Use `fetch-client`. If you need cancellable requests, or IE9 support, use `http-client`.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Add links to the documentation or blog posts.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 29]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $locale
+
+
+#### $location
+
+Use router.navigateToRoute or router.navigate.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Add a link to the documentation.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 30]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $log
+
+Use Aurelia's logger.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Add a link to the documentation.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 31]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $q
+
+No equivalent exists in Aurelia; Aurelia will detect changes made asynchronously. Use ES6 native promises or a promise library.
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: Probably an example of porting `$q.defer()` to `Promise` would be very much appreciated here.

File: doc/article/en-US/angular1-to-aurelia.md [hunk 32]
```diff
@@ -0,0 +1,526 @@
...
+
+#### $templateCache
+
```
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: What is planned here?

**Comment by atsu85** [comment]
@jedd-ahyoung, thanks for the initiative. I undertand it is WIP and in a very early phase as You mentioned, and despite my numerous comments I think it is really great initiative! Thanks for that.

There are some more general comments I want to mention here (not directly related to any specific line of the pull request):
* Mention, that there are way too many ways to write Angular 1 code, but porting Application to aurelia is much easier if application follows suggestions from [Angular styleguide](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md).
* Consider using examples and style of writing Angular code that follow suggestions from [Angular 1 styleguide](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md). I guess sometimes there needs to be more than one example of Angular usage, and how it mapps to Aurelia code.
* Imports are briefly mentioned, but no examples - I think it deserves much more attention.
* `@computedFrom` and [aurelia-computed](http://eisenbergeffect.bluespire.com/aurelia-computed/) plugin are worth mentioning.
* Maybe add more information about `<compose>` or at least mention, that there is more to it, and add a link to documentation.
* Something needs to be discussed for different types of tests as well.

File: doc/article/en-US/angular1-to-aurelia.md (hunk 4 above)
Comment (zewa666) [inline doc/article/en-US/angular1-to-aurelia.md]: I guess its ok since what I understand from the sentence is that you can either setup the router inside the **startup** definition or in the respective viewmodels, e.g. if you use child-routers etc.

File: doc/article/en-US/angular1-to-aurelia.md (hunk 8 above)
Comment (zewa666) [inline doc/article/en-US/angular1-to-aurelia.md]: guess there's an additional type, its not called viewmodel but controller for angular.

File: doc/article/en-US/angular1-to-aurelia.md (hunk 14 above)
Comment (zewa666) [inline doc/article/en-US/angular1-to-aurelia.md]: What a bout providing attaching the code for a gulp task which would automatically modify all templates and switch bindings and interpolations from Angular to Aurelia?

File: doc/article/en-US/angular1-to-aurelia.md (hunk 4 above)
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: oh, You are correct, somehow i didn't pay attention to "Alternatively, " that comes before the rest of the sentence. **I suggest changing "router is configured" to "router could be configured"**.

File: doc/article/en-US/angular1-to-aurelia.md (hunk 14 above)
Comment (atsu85) [inline doc/article/en-US/angular1-to-aurelia.md]: that would be cool :+1:

**Comment by EisenbergEffect** [comment]
Initial version merged. The community can no make PRs to add to or improve the doc.
